records/2021-11-30.performance-testing.md

```diff
@@ -0,0 +1,127 @@
+
+  - Identifying hindrances and impediments with Notify against varying workload levels.
+
+  - Measuring performance temperament of Notify against varying configurations.
+
+  - Discover, test and conclude on scaling strategies for Notify.
+
+- **Expedite capacity planning:**
+
+  - Assist with the deciding on the type of hardware and software resources that are required to run Notify to support pre-defined performance goals.
+
+  - Assist measuring business expectations against the periodic fluctuations of usage, especially as it concerns cost of running the hardware and software infrastructure.
+
+**Requirements:**
+
+* Test the performance of Notify along the following lines:
+  * Notify Endpoint (Response time SLA, Acceptable error rates and Max connection)
+  * Throughput (Change Event to Change Request delay)
+  * Database (Connection on the database, are we overloading the database?)
+
+## Considerations
+
+### Locust Performance Testing Tool
+
+Locust is a scriptable and scalable performance testing tool[[2]](http://docs.locust.io/en/stable/what-is-locust.html).
+
+It requires definition of the performance expectation in regular Python code, and can be executed and presented via UI or headless via a commandline terminal, making it flexible and less restritive to domain specific language[[2]](http://docs.locust.io/en/stable/what-is-locust.html).
+
+This makes Locust infinitely expandable and very developer friendly[[2]](http://docs.locust.io/en/stable/what-is-locust.html).
+
+**Pros:**
+
+* Python based tool which is the main programming language for this project.
+* Notify Team know Python and is used for this project.
+* Locust is hackable allowing the tracking of any Python command or external process.
+* It is extenable via plugins.
+
+**Cons:**
+
+* Have to use external projects to create improved graphs.
+* Runs on a single thread so may have to use multiple worker instances to get high requests per second.
+
+### Infrastructure 
+
+#### Lambda Function
+
+The SRE team initially suggested that we use Lambda Function as the main entry point for running Notify performance test. 
+
+AWS Lambda functions allow code to be run in a highly scalable way. They can integrate properly with AWS products and in this case we are required to run the performance tests once every day after which the results (HTML and CSV) are published to S3. This option was explored because: 
+
+- It has less complexity for the Notify team to manage, the code was lesser compared to ECS.
+- Costs lesser than ECS
+
+We initially went with this option, but soon found out that the file descriptors were capped and the performance tests were quickly running out of connections
+
+> POST,/v2/notifications/email,
+"HTTPSConnectionPool(host='api.staging.notification.cdssandbox.xyz', port=443): 
+Max retries exceeded with url: /v2/notifications/email 
+(Caused by SSLError(OSError(24, 'No file descriptors available')))",1
+
+
+#### Elastic Container Service(ECS)
```

Comment:
Crappy diagram but should be fine for now:
![image](https://user-images.githubusercontent.com/805567/144504454-868f4e37-acb2-41da-aec7-7ba50dd0a1e2.png)

Feel free to edit it [with this link]([![](https://mermaid.ink/img/eyJjb2RlIjoiZ3JhcGggTFJcbiAgICBzdWJncmFwaCBBV1NcbiAgICAgICAgRUJbRXZlbnRCcmlkZ2UgQ3JvbiBTY2hlZHVsZXJdIC0tPnx0cmlnZ2Vyc3wgRUNTKEVDUyBUYXNrKVxuICAgICAgICBzdWJncmFwaCBGYXJnYXRlIENsdXN0ZXJcbiAgICAgICAgRUNTLS0-TFQoTG9jdXN0IFRlc3RzKVxuICAgICAgICBlbmRcbiAgICAgICAgRUNSW0RvY2tlciBpbWFnZSBpbiBFQ1JdLS0-RUNTXG4gICAgICAgIEVDUy0tPlMzW1MzIFN0YWdpbmcgQnVja2V0XVxuICAgIGVuZFxuICAgIHN1YmdyYXBoIEdpdGh1YlxuICAgICAgICBXRltXb3JrZmxvdyBTY3JpcHRdLS0-UzNcbiAgICAgICAgV0YtLT5SRVBPW25vdGlmaWNhdGlvbi1wZXJmb3JtYW5jZS10ZXN0LXJlc3VsdHMgcmVwb3NpdG9yeV1cbiAgICBlbmRcbiIsIm1lcm1haWQiOnsidGhlbWUiOiJkYXJrIn0sInVwZGF0ZUVkaXRvciI6ZmFsc2UsImF1dG9TeW5jIjp0cnVlLCJ1cGRhdGVEaWFncmFtIjpmYWxzZX0)](https://mermaid.live/edit#eyJjb2RlIjoiZ3JhcGggTFJcbiAgICBzdWJncmFwaCBBV1NcbiAgICAgICAgRUJbRXZlbnRCcmlkZ2UgQ3JvbiBTY2hlZHVsZXJdIC0tPnx0cmlnZ2Vyc3wgRUNTKEVDUyBUYXNrKVxuICAgICAgICBzdWJncmFwaCBGYXJnYXRlIENsdXN0ZXJcbiAgICAgICAgRUNTLS0-TFQoTG9jdXN0IFRlc3RzKVxuICAgICAgICBlbmRcbiAgICAgICAgRUNSW0RvY2tlciBpbWFnZSBpbiBFQ1JdLS0-RUNTXG4gICAgICAgIEVDUy0tPlMzW1MzIFN0YWdpbmcgQnVja2V0XVxuICAgIGVuZFxuICAgIHN1YmdyYXBoIEdpdGh1YlxuICAgICAgICBXRltXb3JrZmxvdyBTY3JpcHRdLS0-UzNcbiAgICAgICAgV0YtLT5SRVBPW25vdGlmaWNhdGlvbi1wZXJmb3JtYW5jZS10ZXN0LXJlc3VsdHMgcmVwb3NpdG9yeV1cbiAgICBlbmRcbiIsIm1lcm1haWQiOiJ7XG4gIFwidGhlbWVcIjogXCJkYXJrXCJcbn0iLCJ1cGRhdGVFZGl0b3IiOmZhbHNlLCJhdXRvU3luYyI6dHJ1ZSwidXBkYXRlRGlhZ3JhbSI6ZmFsc2V9)) and export it again.

Mermaid JS code (in case the link does not work):
```
graph LR
    subgraph AWS
        EB[EventBridge Cron Scheduler] -->|triggers| ECS(ECS Task)
        subgraph Fargate Cluster
        ECS-->LT(Locust Tests)
        end
        ECR[Docker image in ECR]-->ECS
        ECS-->S3[S3 Staging Bucket]
    end
    subgraph Github
        WF[Workflow Script]-->S3
        WF-->REPO[notification-performance-test-results repository]
    end
```

Reply:
Really like Mermaid.. will adopt it